nullaway/src/main/java/com/uber/nullaway/handlers/AbstractFieldContractHandler.java

```diff
@@ -0,0 +1,280 @@
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Symbol;
+import com.uber.nullaway.ErrorMessage;
+import com.uber.nullaway.NullAway;
+import com.uber.nullaway.NullabilityUtil;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.AnnotationValue;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.TypeElement;
+
+/**
+ * Abstract base class for handlers that process pre- and post-condition annotations for fields.
+ * Note: all fields that are going to be processed must be the fields of the receiver. (e.g. field
+ * or this.field)
+ */
+public abstract class AbstractFieldContractHandler extends BaseNoOpHandler {
+
+  protected static final String THIS_NOTATION = "this.";
+  /** Simple name of the annotation in {@code String} */
+  protected String annotName;
+
+  /**
+   * Verifies that the processing method adheres to the annotation specifications.
+   *
+   * @param analysis NullAway instance.
+   * @param tree Processing method tree.
+   * @param state Javac {@link VisitorState}.
+   * @param methodSymbol Processing method symbol.
+   */
+  @Override
+  public void onMatchMethod(
+      NullAway analysis, MethodTree tree, VisitorState state, Symbol.MethodSymbol methodSymbol) {
+    Set<String> annotationContent = getFieldNamesFromAnnotation(methodSymbol);
+    boolean isAnnotated = annotationContent != null;
+    boolean isValid =
+        isAnnotated
+            && validateAnnotationSyntax(annotationContent, analysis, tree, state, methodSymbol)
+            && validateAnnotationSemantics(analysis, state, tree, methodSymbol);
+    if (isAnnotated && !isValid) {
+      return;
+    }
+    Symbol.MethodSymbol closestOverriddenMethod =
+        NullabilityUtil.getClosestOverriddenMethod(methodSymbol, state.getTypes());
+    if (closestOverriddenMethod == null) {
+      return;
+    }
+    Set<String> fieldNames;
+    if (isAnnotated) {
+      fieldNames =
+          annotationContent
+              .stream()
+              .map(AbstractFieldContractHandler::trimReceiver)
+              .collect(Collectors.toSet());
+    } else {
+      fieldNames = Collections.emptySet();
+    }
+    validateOverridingRules(fieldNames, analysis, state, tree, closestOverriddenMethod);
+    super.onMatchMethod(analysis, tree, state, methodSymbol);
+  }
+
+  /** Validates whether the annotation conforms to the inheritance rules. */
```

Comment:
Technically this validates whether the `fieldNames` are ok according to the overriding rules.  This method can be called even when a method does not have an annotation.  Fix the comment?